### ISSUE TITLE
Issue in GCM Pusher: failing to call Task.await

### ISSUE THREAD

**Issue by mdillavou** [codedge-llc/pigeon#62]
I have a GenServer running that sends out push notifications using pigeon. I have recently updated to 0.10.3 and I started seeing the following messages every time I sent a GCM push notification:

```
22:51:05.504 [warn] Reverb.Backend.Pusher Reverb.Backend.Pusher received unexpected message in handle_info/2: {#Reference<0.0.3.2852>, [nil]}

22:51:05.504 [warn] Reverb.Backend.Pusher Reverb.Backend.Pusher received unexpected message in handle_info/2: {:DOWN, #Reference<0.0.3.2852>, :process, #PID<0.2565.0>, :normal}
```

After tracking this down, it appears that on line [59 of gcm.ex](https://github.com/codedge-llc/pigeon/blob/master/lib/pigeon/gcm.ex#L59), you are calling `Task.async` but never calling `Task.await`. 

When the Task finishes, this causes the DOWN message to be sent to the parent, which is my GenServer. According to the [elixir documentation](https://hexdocs.pm/elixir/Task.html#async/3), you should always be calling Task.await for every Task.async

**Comment by mkaszubowski** [comment]
I don't want to talk about the implementation, but you can add

```elixir
def handle_info(_, state) do
  {:noreply, state}
end
```
to your GenServer to prevent it from crashing when it receives an unexpected message (I think it's a good practice to always add that clause to your servers). 

Actually, the `{:DOWN, ...}` message can be useful. You can use it if you'd like to retry sending the push if the exit reason is other then `:normal`.

**Comment by mdillavou** [comment]
I have done that, so our GenServer is not crashing. But to me, this behaviour was unexpected. 

I don't think that receiving the DOWN message is that useful. Because we have no knowledge of the Tasks being created, it would be difficult for us to reference each pid and actually understand if something failed or not and why. Your notes say that true asynchronous operation for GCM is coming in the future, so I imagine that implementation will likely resolve this in a more ideal way.

**Comment by hpopp** [comment]
All GCM functionality is about to get redone once #52 is finalized

**Comment by hpopp** [comment]
Updated the PR fixes to use `Task.start_link/1` instead of `spawn`, which appears to be the proper alternative to `Task.async` with no-returns.

I'm going to post the update as `v0.11` because I'm not sure if callback behavior has changed. I needed to tweak my unit tests slightly to make things pass.

**Comment by mdillavou** [comment]
Instead of using Task.start_link directly, you probably want to use it as part of a supervisor tree. I put up a pull request showing how this is done:

https://github.com/codedge-llc/pigeon/pull/64

**Comment by hpopp** [comment]
Yes, you're right. I rarely use `Task` and misread the docs while skimming.

I guess the question then, what are the benefits over a regular `spawn/1`? I noticed the default supervision is `:temporary`

**Comment by mdillavou** [comment]
That's a good question. I'm not sure.

**Comment by talklittle** [comment]
I guess it's better for debugging & inspection purposes, and to fit Erlang/OTP convention, if all your processes are part of the supervision tree. So I think `Task.Supervisor` is actually better for that reason than plain `spawn/1`.

Functionality-wise, they'd behave more-or-less identically.

**Comment by mkaszubowski** [comment]
@talklittle @hpopp 
Actually, there is a difference between `spawn` and using supervised tasks. 

If you use `spawn` there are no link or monitor created for the process. That means that you have no information if the process fails. When using the `Task.Supervisor.async_nolink`, the monitor is created so you can watch for the `{:DOWN, ...}` message and handle the potential failure of the child process. 

There's also a difference when it comes to shutting down the system (although I'm not sure if I'm 100% right on this one). When the process is supervised (`Task.Supervisor`), its supervisor can control if the task is finished between shutting it down. `spawn` just creates an independent process so you have no control over its execution. I'm not so sure of the details but the point is that supervision tree is not only about restarting the failed processes, but also about handling the shutdown in a predictable way (that's why `:temporary` really makes sense). 

The topic seems to be discussed more widely here: http://stackoverflow.com/questions/29326161/what-is-the-point-of-having-temporary-children-for-erlang-supervisors


To sum up, I think that using `Task.Supervisor` is the best option here. If you agree, I'd be happy to send a PR :)

**Comment by hpopp** [comment]
Awesome explanation!
